inference-engine/src/inference_engine/ie_core.cpp

```diff
@@ -158,6 +158,148 @@ class Core::Impl : public ICore {
+            {
+                auto it = config.find(CONFIG_KEY(CACHE_DIR));
+                if (it != config.end()) {
+                    _config[it->first] = it->second;
+                    _modelCacheDir = it->second;
+                    _isModelCacheEnabled = true;
+
+                    // remove if plugin does support common IE Core key
+                    if (!checkPluginSupportsKey(core, it->first, deviceName))
+                        config.erase(it);
+                }
+            }
+        }
+
+        bool isModelCacheEnabled() const { return _isModelCacheEnabled; }
+        std::string getModelCacheDir() const { return _modelCacheDir; }
+    };
+
+    class CompiledBlobHeader final {
+        std::string _ie_version;
+
+    public:
+        CompiledBlobHeader() = default;
+
+        explicit CompiledBlobHeader(const std::string & ie_version) :
+            _ie_version(ie_version) {
+        }
+
+        const std::string & getIeVersion() const {
+            return _ie_version;
+        }
+
+        friend std::istream & operator >> (std::istream & stream, CompiledBlobHeader & header) {
+            std::string xmlStr;
+            std::getline(stream, xmlStr);
+
+            std::cerr << "Header " << std::endl;
+            std::cerr << xmlStr << std::endl;
+
+            pugi::xml_document document;
+            pugi::xml_parse_result res = document.load_string(xmlStr.c_str());
+
+            if (res.status != pugi::status_ok) {
+                THROW_IE_EXCEPTION_WITH_STATUS(NETWORK_NOT_READ) << "Error reading compiled blob header";
+            }
+
+            pugi::xml_node compiledBlobNode = document.document_element();
+            header._ie_version = XMLParseUtils::GetStrAttr(compiledBlobNode, "ie_version");
+
+            return stream;
+        }
+
+        friend std::ostream & operator << (std::ostream & stream, const CompiledBlobHeader & header) {
+            pugi::xml_document document;
+            auto compiledBlobNode = document.append_child("compiled_blob");
+            compiledBlobNode.append_attribute("ie_version").set_value(header._ie_version.c_str());
+
+            document.save(stream, nullptr, pugi::format_raw);
+            document.reset();
+            stream << std::endl;
+
+            return stream;
+        }
+    };
+
+    class CacheManager {
+    public:
+        using Ptr = std::shared_ptr<CacheManager>;
+        virtual ~CacheManager() = default;
+
+        virtual std::shared_ptr<std::ostream> writeCacheEntry(const std::string & id) = 0;
+        virtual std::shared_ptr<std::istream> readCacheEntry(const std::string & id) = 0;
+        virtual void removeCacheEntry(const std::string & id) = 0;
+    };
+
+    class FileStorageCacheManager final : public CacheManager {
+        std::string m_modelCacheDir;
+
+        std::string getBlobFile(const std::string & blobHash) const {
+            return FileUtils::makePath(m_modelCacheDir, blobHash + ".blob");
```

Comment:
How to detect that some existing cache is not valid in this case? My understanding was that hash shall be written to beginning of the file, not represent a file name. And file name shall be calculated based only on CNNNetwork's name and deviceName, like "Resnet50_CPU.blob". So when hash is not matched - old "ResNet50_CPU.blob" will be replaced with new one

Reply:
> And file name shall be calculated based only on CNNNetwork's name and deviceName, like "Resnet50_CPU.blob"

What if I load multiple models with the same name or the same model with different configurations (compile options, different GPUs and so on)? In this case `Resnet50_CPU.blob` will be rewritten multiple times and has no effect.